### PR TITLE
Improved aircraft reloading

### LINKED ISSUE #5966: Implement queued reloading for aircraft [$69]
Since aircraft need to be reloaded, it is very time consuming and boring to reload 20 helicopters at one helipad for example. But if there was a feature that reloaded each aircraft one after another when a group of them are ordered to reload, it would make this process much easier. The same logic could also be applied to repairs at the service depot.

<bountysource-plugin>
---
There is a **[$69 open bounty](https://www.bountysource.com/issues/3148314-implement-queued-reloading-for-aircraft?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F36085&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by bateramos**
Closes #5966

### CODE REVIEW

**Comment by GraionDilach** [comment]
Uhmmm.... uhm.

Contributions are very welcome but you should talk on IRC regarding how to send them. The two patches should have separated into separate PRs and you can check the code style violations locally via make check.

**Comment by Mailaender** [comment]
This is easily repaired with https://help.github.com/articles/using-git-rebase/ though.

**Comment by bateramos** [comment]
yes I know, make check on my computer don't work. I have a MacOS. I try to make to separated then but i failed. Sorry, I will see how can I do this.

**Comment by Mailaender** [comment]
This seems to include the changes from https://github.com/OpenRA/OpenRA/pull/10678 which should be merged first then.

**Comment by GraionDilach** [comment]
@pchote or @Phrohdoh can help you on IRC regarding your Mac issues as both of them are Mac users.

**Comment by bateramos** [comment]
sorry for the confusion people. Its all right now?

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 1]
```diff
@@ -17,6 +17,7 @@ namespace OpenRA.Mods.Common.Activities
 {
 	public class HeliReturnToBase : Activity
 	{
+		const int DistanceAroundHelipadToWaitForTurn = 3000;
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: This...

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 2]
```diff
@@ -46,13 +47,27 @@ public override Activity Tick(Actor self)
...
-				if (nearestHpad == null)
+				if (nearestHpad == null) {
 					return ActivityUtils.SequenceActivities(new Turn(self, initialFacing), new HeliLand(self, true), NextActivity);
-				else
-					return ActivityUtils.SequenceActivities(new HeliFly(self, Target.FromActor(nearestHpad)));
+				} else {
+					var distanceFromHelipad = (nearestHpad.CenterPosition - self.CenterPosition).HorizontalLength;
+
+					//// if no pad is avaliable, move near one and wait
+					if (distanceFromHelipad > 6000) {
```
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: ...and this arbitrary value should be exposed as aircraft properties. In addition, 1 cell = 1024 WDist, so these defaults should preferably be multiples of 1024 (3072 and 6144 respectively).

**Comment by pchote** [comment]
Please note from our [coding standard](https://github.com/OpenRA/OpenRA/wiki/Coding-Standard):
> Braces around code blocks should reside on their own line, and should be omitted if the block contains a single statement (applies to each block separately for if-else statements). The exception to this rule is when there is a short statement (such as a property definition) where the entire statement can reside on a single line.

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs (hunk 2 above)
Comment (bateramos) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: It really should be an aircraft property? Is that this property is valid for just this Activity and, from my tests so far, I did not see any change needed for the different types of units that use this Activity. Maybe expose this constante when there is a need to customize the value?
The value looks kind of arbitrary indeed. With testing with multiple helicopters and only 2 helipads I saw that if they try to place they selfs 3 cells away from the helipad, they don't ram in to each other. 
I will change to 3072. But what do you think about exposing this constante? Do you think its the way to go?

Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: I think making properties accessable is always a good idea, you never know which behaviour a modder wants...

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 3]
```diff
@@ -46,13 +46,28 @@ public override Activity Tick(Actor self)
 									.Where(a => a.Owner == self.Owner && rearmBuildings.Contains(a.Info.Name))
 									.ClosestTo(self);
 
-				if (nearestHpad == null)
+				if (nearestHpad == null) {
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: You can still omit the braces here, so the entire block gets to:
```
if (nearestHpad == null)
	return ActivityUtils.SequenceActivities(...);
else
{
	...
}

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs [hunk 4]
```diff
@@ -46,13 +46,28 @@ public override Activity Tick(Actor self)
...
 					return ActivityUtils.SequenceActivities(new Turn(self, initialFacing), new HeliLand(self, true), NextActivity);
-				else
-					return ActivityUtils.SequenceActivities(new HeliFly(self, Target.FromActor(nearestHpad)));
+				} else {
+					var distanceFromHelipad = (nearestHpad.CenterPosition - self.CenterPosition).HorizontalLength;
+					var distanceLenght = heli.Info.WaitDistanceFromHelipad.Length;
+
+					// If no pad is avaliable, move near one and wait
+					if (distanceFromHelipad > distanceLenght) {
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: Please put the brace on a new line:
```
if (distanceFromHelipad > distanceLenght)
{

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 5]
```diff
@@ -215,6 +228,18 @@ public virtual WVec GetRepulsionForce()
 			return dot >= 0 ? repulsionForce : WVec.Zero;
 		}
 
+		public void DisableRepulsing() {
+			if (Info.Repulsable) {
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: The samehere and below: Put the brace of the method on a new line and remove the braces around the single statement.

File: OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs (hunk 4 above)
Comment (reaperrr) [inline OpenRA.Mods.Common/Activities/Air/HeliReturnToBase.cs]: While you're at it, `Lenght` should be `Length`, and `available` is mistyped as well.

**Comment by abcdefg30** [comment]
Doesn't look too bad otherwise, I just noticed two things:
- Helicopters are still fighting over who goes in first (not sure if we want to fix that in this pr though)
- The descending helicopter will overlap with the next helicopter that is moving towards the pad now, but that's just polish, too

**Comment by bateramos** [comment]
Hello @abcdefg30 
My main focus was fix the resupply of multiple helicopters. I try this approach of just parking the heli near the pad for them to wait for its turn but if another pad just was made available, the waiting heli can use it.
About the first item, that the helis are fighting, actually they can fight for a spot to wait, but reservation of the resupply pad is still working.
Yes you are right about the overlap, but I really did not find a viable solution for this problem. I disabled the repulsion for the descending to fix a problem that a heli can pull a another heli from the pad in a way that it lands on the ground and don't get resupply.
And guys, i'm sorry about the coding style problems, I'm really trying to adjust but is very different from what I was use to it. In this next commit I will revise all code to see if is something else wrong.

**Comment by abcdefg30** [comment]
Well, in this case. :+1: / :white_check_mark: This looks good to me now.

But I'm not sure if it (completely) fixes #5966 as this is only for helicopters.

File: OpenRA.Mods.Common/Traits/Air/Aircraft.cs [hunk 6]
```diff
@@ -215,6 +228,16 @@ public virtual WVec GetRepulsionForce()
 			return dot >= 0 ? repulsionForce : WVec.Zero;
 		}
 
+		public void DisableRepulsing() {
```
Comment (abcdefg30) [inline OpenRA.Mods.Common/Traits/Air/Aircraft.cs]: Oh wait. You still missed to put these braces on a new line.
```
public void DisableRepulsing()
{
```
please. Same for `EnableRepulsing` below.

**Comment by reaperrr** [comment]
@bateramos Are you still working on this? Like @abcdefg30 mentioned, those braces still need to be put on new lines.

**Comment by bateramos** [comment]
Any other code style problem or other issue with the code, I will try to solve as quickly as possible this time.

**Comment by bateramos** [comment]
My fix was on the way that helicopters and helipads behave. What should be the desired behavior of the airports to handle the resupply of multiple planes?

**Comment by abcdefg30** [comment]
Something similiar, I'd guess, as your method for helis works quite well: Planes circle in a waiting position when no airfield is available and then land onto one when it gets free.

**Comment by bateramos** [comment]
Hmmmm ok. About the plane that just resupply and is leaving to make the airport available. He takes flight but where he should wait to receive orders? The helipads have way points, the airports have too?

**Comment by GraionDilach** [comment]
Yes, airports have too.

**Comment by bateramos** [comment]
ok... I think i really need help to solve this one.

**Comment by reaperrr** [comment]
@bateramos 

- First, if you haven't already, make a backup of your local branch of this feature, just in case something goes wrong.
- Next, fetch and pull the latest OpenRA/bleed.
- After that, switch to your local aircraft reloading branch, use git commandline to `git rebase -i HEAD~2` and there, delete the metrics commit, so that only your own commit remains.
- Finally, run `git rebase bleed`, and fix any potential rebase conflicts.

If you followed these steps, your local branch should be ready to do
`git push bateramos +upsteram/#5966`.

**Pull request closed without merging**